src/common/mfu_util.c

```diff
@@ -107,58 +107,89 @@ void daos_bcast_handle(
+  bool connect_pool,
+  bool create_cont)
 {
     /* TODO: if src daos path and dst daos path are false 
      * skip connecting to daos pool */
 
+    /* assume failure until otherwise */
+    int valid = 0;
+    int rc;
+
     /* have rank 0 connect to the pool and container,
      * we'll then broadcast the handle ids from rank 0 to everyone else */
     if (rank == 0) {
-        d_rank_list_t* svcl = daos_rank_list_parse(svc, ":");
-        if (svcl == NULL) {
-            MFU_ABORT(-1, "Failed to parse DAOS rank list: '%s'", svc);
-        }
+        /* Parse svc and connect to DAOS pool */
+        if (connect_pool) {
+            d_rank_list_t* svcl = daos_rank_list_parse(svc, ":");
+            if (svcl == NULL) {
+                MFU_LOG(MFU_LOG_ERR, "Failed to parse DAOS rank list: '%s'", svc);
+                goto bcast;
+            }
 
-        /* Connect to DAOS pool */
-        daos_pool_info_t pool_info;
-        int rc = daos_pool_connect(pool_uuid, NULL, svcl, DAOS_PC_RW,
-                               poh, &pool_info, NULL);
-        if (rc != 0) {
-            MFU_LOG(MFU_LOG_ERR, "Failed to connect to pool");
+            daos_pool_info_t pool_info;
+            rc = daos_pool_connect(pool_uuid, NULL, svcl, DAOS_PC_RW,
+                    poh, &pool_info, NULL);
+            if (rc != 0) {
+                MFU_LOG(MFU_LOG_ERR, "Failed to connect to pool");
+                goto bcast;
+            }
+            d_rank_list_free(svcl);
         }
-        d_rank_list_free(svcl);
 
-        /* attempt to open the daos container */
+        /* Try to open the container */
         daos_cont_info_t co_info;
         rc = daos_cont_open(*poh, cont_uuid, DAOS_COO_RW, coh, &co_info, NULL);
-
+        
         /* If NOEXIST we create it */
         if (rc != 0) {
-            /* create the container */
-            uuid_t cuuid;
-            rc = dfs_cont_create(*poh, cuuid, NULL, NULL, NULL);
+            if (!create_cont) {
+                MFU_LOG(MFU_LOG_ERR, "Failed to open DFS container");
+                goto bcast;
+            }
+            
+            rc = dfs_cont_create(*poh, cont_uuid, NULL, NULL, NULL);
             if (rc != 0) {
                 MFU_LOG(MFU_LOG_ERR, "Failed to create DFS container");
+                goto bcast;
             }
 
             /* try to open it again */
-            rc = daos_cont_open(*poh, cuuid, DAOS_COO_RW, coh, &co_info, NULL);
+            rc = daos_cont_open(*poh, cont_uuid, DAOS_COO_RW, coh, &co_info, NULL);
             if (rc != 0) {
                 MFU_LOG(MFU_LOG_ERR, "Failed to open DFS container");
+                goto bcast;
             }
         }
+
+        /* everything looks good so far */
+        valid = 1;
     }
 
-    /* broadcast pool and container handles from rank 0 */
+bcast:
+    /* broadcast valid from rank 0 */
+    MPI_Bcast(&valid, 1, MPI_INT, 0, MPI_COMM_WORLD);
+
+    /* return if invalid */
+    if (valid == 0) {
+        return -1;
```

Comment:
The valid check to guard the handle bast looks good.  Thanks.  One other thing to check is whether we should clean up any resources that may have been allocated if we fail.  I don't know the daos api well enough to check, but I did see one call that looked to trying to free something above which would not be called in certain error cases.

Reply:
Yes, you are right. I missed that. Thank you!

Reply:
Luckily, I think only the d_rank_list_free was missing here, since the pool/container handles are declared statically as structs in dcp.c, and the daos calls expect them to not be NULL, so I don't believe the daos calls allocate anything else here.